### PR TITLE
Map func dedupe & fire breath effect

### PR DESCRIPTION
**Pull request by ethteck**
Tried to name stuff in fire breath as much as I reasonably could. pretty happy with it

### CODE REVIEW

**Comment by BowserSlug** [comment]
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀 This PR matches 32 functions (+1.02%) on `us`.

⚠️ This PR introduces 78 warnings:
- src/165490.c:229: warning: implicit declaration of function `filemenu_draw_file_name'
- src/18F340.c:342: warning: assignment from incompatible pointer type
- src/18F340.c:347: warning: assignment from incompatible pointer type
- src/18F340.c:361: warning: assignment from incompatible pointer type
- src/18F340.c:370: warning: assignment from incompatible pointer type
- src/1a1f0_len_5390.c:140: warning: assignment from incompatible pointer type
- src/1a1f0_len_5390.c:145: warning: assignment from incompatible pointer type
- src/1a1f0_len_5390.c:159: warning: assignment from incompatible pointer type
- src/1a1f0_len_5390.c:168: warning: assignment from incompatible pointer type
- src/battle/partner/watt.c:129: warning: assignment from incompatible pointer type
- src/battle/partner/watt.c:90: warning: assignment from incompatible pointer type
- src/effects/attack_result_text.c:53: warning: assignment from incompatible pointer type
- src/effects/attack_result_text.c:92: warning: initialization from incompatible pointer type
- src/effects/aura.c:45: warning: assignment from incompatible pointer type
- src/effects/balloon.c:34: warning: assignment from incompatible pointer type
- src/effects/balloon.c:89: warning: initialization from incompatible pointer type
- src/effects/big_smoke_puff.c:123: warning: initialization from incompatible pointer type
- src/effects/big_smoke_puff.c:48: warning: assignment from incompatible pointer type
- src/effects/big_smoke_puff.c:77: warning: initialization from incompatible pointer type
- src/effects/blast.c:27: warning: assignment from incompatible pointer type
- src/effects/blast.c:51: warning: initialization from incompatible pointer type
- src/effects/cloud_puff.c:117: warning: initialization from incompatible pointer type
- src/effects/cloud_puff.c:30: warning: assignment from incompatible pointer type
- src/effects/cloud_trail.c:118: warning: initialization from incompatible pointer type
- src/effects/cloud_trail.c:29: warning: assignment from incompatible pointer type
- src/effects/debuff.c:135: warning: initialization from incompatible pointer type
- src/effects/debuff.c:32: warning: assignment from incompatible pointer type
- src/effects/debuff.c:86: warning: initialization from incompatible pointer type
- src/effects/effect_3C.c:18: warning: initialization from incompatible pointer type
- src/effects/effect_75.c:25: warning: initialization from incompatible pointer type
- src/effects/effect_86.c:112: warning: initialization from incompatible pointer type
- src/effects/effect_86.c:48: warning: assignment from incompatible pointer type
- src/effects/effect_86.c:79: warning: initialization from incompatible pointer type
- src/effects/emote.c:118: warning: initialization from incompatible pointer type
- src/effects/emote.c:77: warning: assignment from incompatible pointer type
- src/effects/ending_decals.c:18: warning: initialization from incompatible pointer type
- src/effects/energy_orb_wave.c:18: warning: initialization from incompatible pointer type
- src/effects/fire_flower.c:30: warning: assignment from incompatible pointer type
- src/effects/flame.c:27: warning: initialization from incompatible pointer type
- src/effects/floating_flower.c:146: warning: initialization from incompatible pointer type
- src/effects/floating_flower.c:29: warning: assignment from incompatible pointer type
- src/effects/floating_flower.c:79: warning: initialization from incompatible pointer type
- src/effects/floating_rock.c:19: warning: initialization from incompatible pointer type
- src/effects/flower_splash.c:142: warning: initialization from incompatible pointer type
- src/effects/flower_splash.c:65: warning: assignment from incompatible pointer type
- src/effects/flower_trail.c:160: warning: initialization from incompatible pointer type
- src/effects/flower_trail.c:67: warning: assignment from incompatible pointer type
- src/effects/footprint.c:119: warning: initialization from incompatible pointer type
- src/effects/footprint.c:41: warning: assignment from incompatible pointer type
- src/effects/landing_dust.c:108: warning: assignment from incompatible pointer type
- src/effects/landing_dust.c:216: warning: initialization from incompatible pointer type
- src/effects/landing_dust.c:261: warning: initialization from incompatible pointer type
- src/effects/lightning_bolt.c:29: warning: initialization from incompatible pointer type
- src/effects/music_note.c:53: warning: assignment from incompatible pointer type
- src/effects/music_note.c:94: warning: initialization from incompatible pointer type
- src/effects/shape_spell.c:104: warning: initialization from incompatible pointer type
- src/effects/shape_spell.c:42: warning: assignment from incompatible pointer type
- src/effects/shape_spell.c:79: warning: initialization from incompatible pointer type
- src/effects/sleep_bubble.c:27: warning: assignment from incompatible pointer type
- src/effects/sleep_bubble.c:49: warning: initialization from incompatible pointer type
- src/effects/small_gold_sparkle.c:113: warning: initialization from incompatible pointer type
- src/effects/small_gold_sparkle.c:45: warning: assignment from incompatible pointer type
- src/effects/snowflake.c:123: warning: initialization from incompatible pointer type
- src/effects/snowflake.c:30: warning: assignment from incompatible pointer type
- src/effects/snowflake.c:61: warning: initialization from incompatible pointer type
- src/effects/snowflake.c:86: warning: initialization from incompatible pointer type
- src/effects/star.c:120: warning: initialization from incompatible pointer type
- src/effects/star.c:182: warning: initialization from incompatible pointer type
- src/effects/star.c:43: warning: assignment from incompatible pointer type
- src/effects/stat_change.c:151: warning: assignment from incompatible pointer type
- src/effects/stat_change.c:180: warning: initialization from incompatible pointer type
- src/effects/tattle_window.c:107: warning: initialization from incompatible pointer type
- src/effects/tattle_window.c:75: warning: assignment from incompatible pointer type
- src/effects/walking_dust.c:39: warning: assignment from incompatible pointer type
- src/effects/walking_dust.c:97: warning: initialization from incompatible pointer type
- src/npc.c:1051: warning: assignment from incompatible pointer type
- src/npc.c:1071: warning: assignment from incompatible pointer type
- src/npc.c:1094: warning: assignment from incompatible pointer type
- src/npc.c:999: warning: assignment from incompatible pointer type

File: include/effects.h [hunk 1]
```diff
@@ -734,6 +760,12 @@ typedef struct Effect6FInstance {
     /* 0x10 */ struct EffectGraphics* effect;
 } Effect6FInstance;
 
+enum FireBreathType {
+    FIRE_BREATH_LARGE = 0,
+    FIRE_BREATH_SMALL = 1,
+    FIRE_BREATH_TINY = 2,
```
Comment (nanaian) [inline include/effects.h:766]:
```suggestion
    FIRE_BREATH_TINY  = 2,
```

File: src/effects/fire_breath.c [hunk 2]
```diff
@@ -1,30 +1,171 @@
...
+
+void fire_breath_init(EffectInstance* effect);
+void fire_breath_update(EffectInstance* effect);
+void fire_breath_render(EffectInstance* effect);
 void fire_breath_appendGfx(void* effect);
 
-INCLUDE_ASM(s32, "effects/fire_breath", fire_breath_main);
+EffectInstance* fire_breath_main(s32 type, f32 startX, f32 startY, f32 startZ, f32 endX, f32 endY, f32 endZ,
+                                 s32 numExtra, s32 spawnDelay, s32 lifetime)
+{
```
Comment (nanaian) [inline src/effects/fire_breath.c:19]:
```suggestion
EffectInstance* fire_breath_main(
    s32 type,
    f32 startX, f32 startY, f32 startZ,
    f32 endX, f32 endY, f32 endZ,
    s32 numExtra, s32 spawnDelay, s32 lifetime
) {
```

File: src/effects/fire_breath.c [hunk 3]
```diff
@@ -1,30 +1,171 @@
...
+    maxLifetime = data->maxLifetime;
+    spawnTimer = data->spawnTimer;
+
+    if (lifetime >= 6 && data->type == FIRE_BREATH_LARGE) {
+        data->scale += (2.5 - data->scale) * 0.05;
+    }
+
+    data->pos.x = data->initPos.x + ((((data->endPos.x - data->initPos.x) + data->unk_50.x) * spawnTimer) / maxLifetime);
+    data->pos.y = data->initPos.y + ((((data->endPos.y - data->initPos.y) + data->unk_50.y) * spawnTimer) / maxLifetime);
+    data->pos.z = data->initPos.z + ((((data->endPos.z - data->initPos.z) + data->unk_50.z) * spawnTimer) / maxLifetime);
```
Comment (nanaian) [inline src/effects/fire_breath.c:116]:
```suggestion
    data->pos.x = data->initPos.x + (((data->endPos.x - data->initPos.x + data->unk_50.x) * spawnTimer) / maxLifetime);
    data->pos.y = data->initPos.y + (((data->endPos.y - data->initPos.y + data->unk_50.y) * spawnTimer) / maxLifetime);
    data->pos.z = data->initPos.z + (((data->endPos.z - data->initPos.z + data->unk_50.z) * spawnTimer) / maxLifetime);
```

File: src/world/common/UnkFunc55.inc.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        case 1:
+            x = npc->pos.x;
+            y = npc->pos.y;
+            z = npc->pos.z;
+            if (npc_test_move_simple_with_slipping(0, &x, &y, &z, npc->moveSpeed, npc->yaw, npc->collisionRadius,
+                npc->collisionHeight) == 0)
+            {
+                npc_move_heading(npc, npc->moveSpeed, npc->yaw);
+            } else {
+
```
Comment (nanaian) [inline src/world/common/UnkFunc55.inc.c:63]:
```suggestion
```

**Review by nanaian — COMMENTED** [review]
style stuff

**Pull request merged**